proj2aci/asset.go

```diff
@@ -0,0 +1,259 @@
+	return newPath
+}
+
+func validateAsset(ACIAsset, localAsset string) error {
+	if !filepath.IsAbs(ACIAsset) {
+		return fmt.Errorf("Wrong ACI asset: '%v' - ACI asset has to be absolute path", ACIAsset)
+	}
+	if !filepath.IsAbs(localAsset) {
+		return fmt.Errorf("Wrong local asset: '%v' - local asset has to be absolute path", localAsset)
+	}
+	fi, err := os.Lstat(localAsset)
+	if err != nil {
+		return fmt.Errorf("Error stating %v: %v", localAsset, err)
+	}
+	mode := fi.Mode()
+	if mode.IsDir() || mode.IsRegular() || isSymlink(mode) {
+		return nil
+	}
+	return fmt.Errorf("Can't handle local asset %v - not a file, not a dir, not a symlink", fi.Name())
+}
+
+func copyTree(src, dest string) error {
+	return filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		rootLess := path[len(src):]
+		target := filepath.Join(dest, rootLess)
+		mode := info.Mode()
+		switch {
+		case mode.IsDir():
+			err := os.Mkdir(target, mode.Perm())
+			if err != nil {
+				return err
+			}
+		case mode.IsRegular():
+			if err := copyRegularFile(path, target); err != nil {
+				return err
+			}
+		case isSymlink(mode):
+			if err := copySymlink(path, target); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("Unsupported node %q in assets, only regular files, directories and symlinks are supported.", path, mode.String())
+		}
+		return nil
+	})
+}
+
+func copyRegularFile(src, dest string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+	destFile, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer destFile.Close()
+	if _, err := io.Copy(destFile, srcFile); err != nil {
+		return err
+	}
+	fi, err := srcFile.Stat()
+	if err != nil {
+		return err
+	}
+	if err := destFile.Chmod(fi.Mode().Perm()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func copySymlink(src, dest string) error {
+	symTarget, err := os.Readlink(src)
+	if err != nil {
+		return err
+	}
+	if err := os.Symlink(symTarget, dest); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getSoLibs tries to run ldd on given path and to process its output
+// to get a list of shared libraries to copy. This list is returned as
+// an array of assets.
+//
+// man ldd says that running ldd on untrusted executables is dangerous
+// (it might run an executable to get the libraries), so possibly this
+// should be replaced with objdump. Problem with objdump is that it
+// just gives library names, while ldd absolute paths to those
+// libraries - to use objdump we need to know the $libdir.
+func getSoLibs(path string) ([]string, error) {
+	assets := []string{}
+	buf := new(bytes.Buffer)
+	args := []string{
+		"ldd",
+		path,
+	}
+	if err := RunCmdFull("", args, nil, "", buf, nil); err != nil {
+		if _, ok := err.(CmdFailedError); !ok {
+			return nil, err
+		}
+	} else {
+		re := regexp.MustCompile(`(?m)^\t(?:\S+\s+=>\s+)?(\S+)\s+\([0-9a-fA-Fx]+\)$`)
```

Comment:
My ldd output is as follows:

```
ldd /bin/ls
        linux-vdso.so.1 (0x00007ffe6f3f5000)
[...]
```

So this regexp will match linux-vdso.so.1 which is not an absolute path and it will fail.